### PR TITLE
IR generation for constants.

### LINKED ISSUE #8344: [Sol - Yul] Support constants
```
contract Hash {
    bytes32 constant hash = hex"1234";

    function f() public pure returns(bytes32) {
        return hash;
    }
}
```

### PR DESCRIPTION
**Pull request by chriseth**
Closes #8344.

### CODE REVIEW

**Comment by axic** [comment]
Rebased.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `baedc1b2d093fb823d0063e56179f56595b9fab8`:
```
Coding style error:
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1417: define(_referencingExpression) << constantValueFunction(_variable) << "()\n";

```
Please check that your changes are working as intended.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -1412,12 +1412,10 @@ void IRGeneratorForStatements::handleVariableReference(
...
 )
 {
-	// TODO for the constant case, we have to be careful:
-	// If the value is visited twice, `defineExpression` is called twice on
-	// the same expression.
-	solUnimplementedAssert(!_variable.isConstant(), "");
 	solUnimplementedAssert(!_variable.immutable(), "");
-	if (m_context.isLocalVariable(_variable))
+	if (_variable.isStateVariable() && _variable.isConstant())
+	    define(_referencingExpression) << constantValueFunction(_variable) << "()\n";
```
Comment (stackenbotten) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Coding style error

File: test/libsolidity/semanticTests/constants/constant_string.sol [hunk 2]
```diff
@@ -16,7 +16,9 @@ contract C {
     }
 }
 
+// ====
+// compileViaYul: also
 // ----
-// f() -> 0x20, 3, "\x03\x01\x02"
-// g() -> 0x20, 3, "\x03\x01\x02"
+// f() -> 0x20, 3, ""
+// g() -> 0x20, 3, ""
```
Comment (axic) [inline test/libsolidity/semanticTests/constants/constant_string.sol]: Seems to be broken.

Comment (chriseth) [reply test/libsolidity/semanticTests/constants/constant_string.sol]: hm, that seems to have been just the tool

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -1576,11 +1598,9 @@ void IRGeneratorForStatements::handleVariableReference(
 	Expression const& _referencingExpression
 )
 {
-	// TODO for the constant case, we have to be careful:
-	// If the value is visited twice, `defineExpression` is called twice on
-	// the same expression.
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1581]: I'm not quite sure I understand what this old comment warned about...

**Review by ekpyron — DISMISSED** [review]
Looks good - even though I'm not quite sure about that removed comment...

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 4]
```diff
@@ -268,6 +268,17 @@ string IRGenerator::generateGetter(VariableDeclaration const& _varDecl)
...
+			else if (_varDecl.isConstant())
+				return Whiskers(R"(
+					function <functionName>() -> <ret> {
+						<ret> := <constantValueFunction>()
+					}
+				)")
+				("functionName", functionName)
+				("constantValueFunction", IRGeneratorForStatements(m_context, m_utils).constantValueFunction(_varDecl))
+				("ret", suffixedVariableNameList("ret_", 0, _varDecl.type()->sizeOnStack()))
+				("id", to_string(_varDecl.id()))
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGenerator.cpp]: It is not needed, no `<id>` in template?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1581]: It warned that you cannot just generate code for the expression like we did in the old codegen. Instead, I created a function that is called here and that contains the codegen for the expression only once.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp]: fixed.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 5]
```diff
@@ -202,11 +202,11 @@ string IRGenerator::generateGetter(VariableDeclaration const& _varDecl)
 
 	Type const* type = _varDecl.annotation().type;
 
-	solAssert(!_varDecl.isConstant(), "");
 	solAssert(_varDecl.isStateVariable(), "");
 
 	if (auto const* mappingType = dynamic_cast<MappingType const*>(type))
 		return m_context.functionCollector().createFunction(functionName, [&]() {
+			solAssert(!_varDecl.isConstant(), "");
```
Comment (axic) [inline libsolidity/codegen/ir/IRGenerator.cpp:209]: Are immutable mappings supported? If not maybe add another assert here for !immutable.

File: test/libsolidity/semanticTests/constants/constant_string.sol (hunk 2 above)
Comment (axic) [reply test/libsolidity/semanticTests/constants/constant_string.sol]: @mijovic can you check this in the tool and perhaps create a tracking issue (to be fixed later)?

**Review by axic — COMMENTED** [review]
FWIW looks okay to me, but I haven't immersed myself into the internals into the IRGenerator lately.

**Review by ekpyron — APPROVED** [review]

File: test/libsolidity/semanticTests/constants/constant_string.sol (hunk 2 above)
Comment (mijovic) [reply test/libsolidity/semanticTests/constants/constant_string.sol]: I'll check behavior of isoltest here

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/codegen/ir/IRGenerator.cpp:209]: @chriseth @ekpyron there was no answer to this ^^

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp:209]: not sure what an immutable mapping should be, but I can add the assertion.

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp:209]: https://github.com/ethereum/solidity/pull/8871

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerator.cpp:209]: Yeah... I thought about pointing towards this when approving, but then wasn't sure there can even be immutable mappings, so I ignored it :-).

Comment (axic) [reply libsolidity/codegen/ir/IRGenerator.cpp:209]: > not sure what an immutable mapping should be,

The same as other immutable value types (e.g. set in the constructor only)?

File: test/libsolidity/semanticTests/constants/constant_string.sol (hunk 2 above)
Comment (mijovic) [reply test/libsolidity/semanticTests/constants/constant_string.sol]: I can't reproduce this

Comment (mijovic) [reply test/libsolidity/semanticTests/constants/constant_string.sol]: After rebasing I got it. Will investigate and create tracking issue.

Comment (mijovic) [reply test/libsolidity/semanticTests/constants/constant_string.sol]: #8885

**Pull request merged**